### PR TITLE
Json

### PR DESCRIPTION
**Pull request by eszterpos**
This branch contains the implementation of the handling of .json type extras strings.
It includes an implementation of a new data structure (listDict in the new tools/xtratools.py) that enables us to access the extra strings in a smoother way. The object stores a list of dictionaries 

In case the driver does not pass a JSON-formatted string, the mxtra dictionary creates an entry with {'info': string passed by the driver}.

Also, there is a new way of handling the output of extras, and to demonstrate this, I have slightly modified some existing examples. The novelty hides behind the new attribute of the trajectoryOutput object, 'xtratype' which enables the user to print out information to separate files depending on what extras strings have been received and what is of interest.

Moreover, I added a way to promote extras numerical data to be interpolated over in case the bead-contraction scheme is being used.

### CODE REVIEW

File: examples/pes/instanton/ch4h-cbe-rate/instanton/20_non_uniform_discret/no_distr [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#Bead    Energy (eV)
```
Comment (litman90) [inline examples/pes/instanton/ch4h-cbe-rate/instanton/20_non_uniform_discret/no_distr:1]: What is this file for?

File: examples/pes/instanton/ch4h-cbe-rate/instanton/20_non_uniform_discret/input.xml [hunk 2]
```diff
@@ -8,7 +8,7 @@
    </ffsocket>
    <system >
        <initialize nbeads='20'> 
-           <file mode='xyz' >  init.xyz </file>
+           <file mode='xyz' >  in.xyz </file>
```
Comment (litman90) [inline examples/pes/instanton/ch4h-cbe-rate/instanton/20_non_uniform_discret/input.xml:11]: why this file has been modified?
I see the init.xyz so this folder needs some cleaning.

File: ipi/README [hunk 3]
```diff
@@ -0,0 +1,22 @@
+This branch contains the implementation of the .JSON extra files.
```
Comment (litman90) [inline ipi/README:1]: This file should not be here

File: ipi/engine/motion/instanton.py [hunk 4]
```diff
@@ -938,6 +959,39 @@ def pre_step(self, step=None, adaptative=False):
         self.qtime = -time.time()
         info("\n Instanton optimization STEP {}".format(step), verbosity.low)
 
+        # Choosing the forcefield that has returned friction tensor in extras
```
Comment (litman90) [inline ipi/engine/motion/instanton.py:962]: The current PR is about the extras communication, I would delete this here.

File: ipi/inputs/motion/instanton.py [hunk 5]
```diff
@@ -49,18 +49,6 @@ class InputInst(InputDictionary):
         )
     }
```
Comment (litman90) [inline ipi/inputs/motion/instanton.py:51]: Why this inputAtributre has been eliminated?

File: ipi/inputs/outputs.py [hunk 6]
```diff
@@ -166,6 +166,22 @@ class InputTrajectory(InputValue):
...
+        {
+            "dtype": str,
+            "default": "info",
+            "help": "What extra to print from the different extra strings.",
+            "options": [
+                "info",
+                "dipole",
+                "friction",
+                "hirshfeld",
+                "workfunction",
```
Comment (litman90) [inline ipi/inputs/outputs.py:180]: maybe change it to work_function to keep it consistent with atomic_stress?

File: ipi_tests/regression_tests/tests/INSTANTON/200K/input.xml [hunk 7]
```diff
@@ -37,3 +37,4 @@
        </motion>
    </system>
 </simulation>
+
```
Comment (litman90) [inline ipi_tests/regression_tests/tests/INSTANTON/200K/input.xml:40]: please delete extra space here

File: ipi_tests/regression_tests/tests/GEOP/BFGS/pswater/RESTART [hunk 8]
```diff
@@ -0,0 +1,73 @@
+<simulation verbosity='medium' mode='static'>
```
Comment (litman90) [inline ipi_tests/regression_tests/tests/GEOP/BFGS/pswater/RESTART:1]: Please delete this file

File: .gitignore [hunk 9]
```diff
@@ -13,3 +13,4 @@ doc/manual.xml
 \#*
 *.orig
 regtest-ref
+JSONtest
```
Comment (litman90) [inline .gitignore:16]: The gitignore is not intended for specific tests purpose during development. Please delete this line

File: ipi/interfaces/sockets.py [hunk 10]
```diff
@@ -476,6 +486,55 @@ def dispatch(self, r):
         # marks the request as done as the very last thing
         r["status"] = "Done"
 
+    # def getjson(self):
```
Comment (litman90) [inline ipi/interfaces/sockets.py:489]: please delete this

**Comment by litman90** [comment]
Please also add the regression tests required to test the new functionality

File: ipi/inputs/motion/instanton.py (hunk 5 above)
Comment (eszterpos) [reply ipi/inputs/motion/instanton.py:51]: the whole block is completely repeated in the master branch, take a look :)

File: ipi/interfaces/sockets.py (hunk 10 above)
Comment (eszterpos) [reply ipi/interfaces/sockets.py:489]: deleted.

File: ipi_tests/regression_tests/tests/GEOP/BFGS/pswater/RESTART (hunk 8 above)
Comment (eszterpos) [reply ipi_tests/regression_tests/tests/GEOP/BFGS/pswater/RESTART:1]: deleted.

File: ipi_tests/regression_tests/tests/INSTANTON/200K/input.xml (hunk 7 above)
Comment (eszterpos) [reply ipi_tests/regression_tests/tests/INSTANTON/200K/input.xml]: deleted.

File: ipi/inputs/outputs.py (hunk 6 above)
Comment (eszterpos) [reply ipi/inputs/outputs.py:180]: fair point! corrected it :)

File: ipi/engine/motion/instanton.py (hunk 4 above)
Comment (eszterpos) [reply ipi/engine/motion/instanton.py:962]: That is true, however, I think that the point why we are doing the whole modification of the extras input, is to have access to the friction. So I included this part to show how one accesses, the friction how to transfer that data structure into a friction object to work with.

File: ipi/README (hunk 3 above)
Comment (eszterpos) [reply ipi/README]: deleted

File: .gitignore (hunk 9 above)
Comment (eszterpos) [reply .gitignore:16]: yes, I had it there only not to synchronise the directory, I am going to make the regtests from the content of the above directory and will remove it from here, thanks! :)

File: ipi/engine/motion/instanton.py (hunk 4 above)
Comment (mahrossi) [reply ipi/engine/motion/instanton.py:962]: I agree that this can/should be deleted.

**Comment by mahrossi** [comment]
Is this good to go? seems to me. Yair, approve?

File: ipi/engine/motion/instanton.py (hunk 4 above)
Comment (eszterpos) [reply ipi/engine/motion/instanton.py:962]: I understand, yes :) I removed it.

File: .gitignore (hunk 9 above)
Comment (litman90) [reply .gitignore]: is this resolved eszter? if so please press "resolve conversation"

File: ipi/engine/motion/instanton.py (hunk 4 above)
Comment (litman90) [reply ipi/engine/motion/instanton.py:962]: is this resolved eszter? if so please press "resolve conversation"

**Review by eszterpos — COMMENTED** [review]
> Please also add the regression tests required to test the new functionality

Yes, I added 2 tests, these are testing the xtratype functionality. I compared the numbers printed by the driver with the numbers that are being printed by i-pi.

**Comment by litman90** [comment]
Eszter could you explain how you generated the FRICTION_DIPOLE example? I don't get it.

**Comment by litman90** [comment]
I merged this branch with a copy of my working instanton branch to test it. 
The functionality works as expected =)

I have the  following suggestions,
 - add a default value to the new_x argument in 'transfer_forces_manual', so that function can be used without taking care of the presence of extras. A warning is in order though.
 - remove all the modifications to the instanton.py. This PR should not modify that file.
  (I already took all the relevant parts of the code to perform the tests ;) )

**Comment by eszterpos** [comment]
There is a part in the instanton.py where we interpolate for the potential and the forces. I extended that part for the extras numerical inputs. I think this is a logical decision. Can you explain why should we remove that part? I agree that it would be more logical to be done in a different part of the code, so maybe placed somewhere else is better, could you maybe suggest a better place for this in the code?

**Comment by litman90** [comment]
> There is a part in the instanton.py where we interpolate for the potential and the forces. I extended that part for the extras numerical inputs. I think this is a logical decision. Can you explain why should we remove that part? I agree that it would be more logical to be done in a different part of the code, so maybe placed somewhere else is better, could you maybe suggest a better place for this in the code?

That part of the code doesn't exist anymore so it is not possible to say, put it here or there.
This PR makes available the "extra" information but it doesn't care how you use it in other parts of the code (like in this case, an interpolation)

**Comment by eszterpos** [comment]
I added 2 regression tests testing the xtratools infrastructure, and removed the parts of the instanton.py file that are being partially reorganised and rewritten by @litman90 in the instanton 3.0 branch. I also added some fixes for the case when no extras are being passed. 

As the first regtests I created seemed to be unstable under the github CI, I now checked that the new ones are stable on both i-pi main repo and on the sabia repo, too. 

I consider the PR ready to merge from my side. What do you think, @litman90 and @mahrossi ? :)

File: ipi/engine/forces.py [hunk 11]
```diff
@@ -894,35 +895,42 @@ def transfer_forces(self, refforce):
...
           - new_f list of length equal to number of force type, containing the beads forces
+          - new_x list of length equal to number of force type, containing the beads extras
         """
         msg = "Unconsistent dimensions inside transfer_forces_manual"
         assert len(self.mforces) == len(new_q), msg
         assert len(self.mforces) == len(new_v), msg
         assert len(self.mforces) == len(new_forces), msg
+        if new_x == None:
+            new_x = [[None] * self.nbeads] * len(self.mforces)
+            warning("WARNING: No extras information has been passed.")
```
Comment (litman90) [inline ipi/engine/forces.py]: This should not be a warning.

Comment (litman90) [reply ipi/engine/forces.py]: I will downgrade it to an info msg

**Comment by litman90** [comment]
It looks good to be merged after the tests finish without problems

**Review by litman90 — APPROVED** [review]

**Pull request merged**